src/chainstate/stacks/index/storage.rs

```diff
@@ -0,0 +1,1213 @@
+        let nr = self.read_node_count;
+        let br = self.read_backptr_count;
+        let nw = self.write_node_count;
+
+        self.read_node_count = 0;
+        self.read_backptr_count = 0;
+        self.write_node_count = 0;
+         
+        (nr, br, nw)
+    }
+
+    #[cfg(test)]
+    #[allow(dead_code)]
+    pub fn leaf_stats(&mut self) -> (u64, u64) {
+        let lr = self.read_leaf_count;
+        let lw = self.write_leaf_count;
+
+        self.read_leaf_count = 0;
+        self.write_leaf_count = 0;
+
+        (lr, lw)
+    }
+
+    /// Walk through the bufferred TrieNodes and dump them to f.
+    fn dump_traverse<F: Read + Write + Seek>(&mut self, f: &mut F, root: &TrieNodeType, hash: &TrieHash, parent_hash: &BlockHeaderHash) -> Result<u64, Error> {
+        let mut frontier : VecDeque<(TrieNodeType, TrieHash)> = VecDeque::new();
+
+        let mut node_data = vec![];
+        let mut offsets = vec![];
+
+        frontier.push_back((root.clone(), hash.clone()));
+
+        let mut ptr = BLOCK_HEADER_HASH_ENCODED_SIZE as u64;       // first 32 bytes is reserved for the parent block hash
+        
+        // step 1: write out each node in breadth-first order to get their ptr offsets
+        while frontier.len() > 0 {
+            let (node, node_hash) = match frontier.pop_front() {
+                Some((n, h)) => (n, h),
+                None => {
+                    break;
+                }
+            };
+
+            // calculate size
+            let num_written = get_node_byte_len(&node);
+            ptr += num_written as u64;
+            
+            // queue each child
+            if !node.is_leaf() {
+                let ptrs = node.ptrs();
+                let num_children = ptrs.len();
+                for i in 0..num_children {
+                    if ptrs[i].id != TrieNodeID::Empty && !is_backptr(ptrs[i].id) {
+                        let (child, child_hash) = self.read_nodetype(&ptrs[i])?;
+                        frontier.push_back((child, child_hash));
+                    }
+                }
+            }
+            
+            node_data.push((node, node_hash));
+            offsets.push(ptr as u32);
+        }
+
+        assert_eq!(offsets.len(), node_data.len());
+
+        // step 2: update ptrs in all nodes
+        let mut i = 0;
+        for j in 0..node_data.len() {
+            let next_node = &mut node_data[j].0;
+            if !next_node.is_leaf() {
+                let mut ptrs = next_node.ptrs_mut();
+                let num_children = ptrs.len();
+                for k in 0..num_children {
+                    if ptrs[k].id != TrieNodeID::Empty && !is_backptr(ptrs[k].id) {
+                        ptrs[k].ptr = offsets[i];
+                        i += 1;
+                    }
+                }
+            }
+        }
+
+        // step 3: write out each node (now that they have the write ptrs)
+        frontier.push_back((root.clone(), hash.clone()));
```

Comment:
Does frontier need to be re-initialized here?

Reply:
The frontier will have been drained before this code is reached.  But, I can make that explicit with a `.clear()`.

Reply:
But why push a clone of root and hash back onto the frontier?

Reply:
Ah lol you're right.  I think there used to be some code after this that ended up getting removed.

Reply:
Should be fixed now.